### PR TITLE
Enable core limits on performance testing client and server

### PR DESCRIPTION
**Pull request by vjpai**
Command line just specifies the core limit for the client and server
The driver then figures out which cores to give to client and server
If the client and server are running on the same specified host, the driver will make sure that if only one (client or server) limit is specified, that the other one is automatically determined.

### CODE REVIEW

**Comment by vjpai** [comment]
Reassigning to improve reviewer load-balance

File: test/cpp/qps/limit_cores.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#ifndef TEST_QPS_LIMIT_CORES_H
+#define TEST_QPS_LIMIT_CORES_H
+
+#include <vector>
+
+namespace grpc {
+namespace testing {
+// LimitCores takes array and size arguments (instead of vector) for more direct
+// conversion from repeated field of protobuf. Use a cores_size of 0 to remove
+// existing limits (from an empty repeated field)
```
Comment (dgquintas) [inline test/cpp/qps/limit_cores.h]: But what is it for?

File: test/cpp/qps/limit_cores.cc [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+namespace grpc {
+namespace testing {
+
+#ifdef GPR_CPU_LINUX
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+#include <sched.h>
+int LimitCores(const int *cores, int cores_size) {
+  int num_cores = gpr_cpu_num_cores();
```
Comment (dgquintas) [inline test/cpp/qps/limit_cores.cc]: const

File: test/cpp/qps/limit_cores.cc [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  int cores_set = 0;
+
+  cpu_set_t *cpup = CPU_ALLOC(num_cores);
+  GPR_ASSERT(cpup);
+  size_t size = CPU_ALLOC_SIZE(num_cores);
```
Comment (dgquintas) [inline test/cpp/qps/limit_cores.cc]: const

File: test/cpp/qps/driver.cc [hunk 4]
```diff
@@ -59,7 +60,42 @@ using std::vector;
 
 namespace grpc {
 namespace testing {
-static deque<string> get_hosts(const string& name) {
+static std::string get_host(const std::string& worker) {
+  char* host;
+  char* port;
+
+  gpr_split_host_port(worker.c_str(), &host, &port);
+  string s(host);
```
Comment (dgquintas) [inline test/cpp/qps/driver.cc]: const

File: test/cpp/qps/driver.cc [hunk 5]
```diff
@@ -59,7 +60,42 @@ using std::vector;
...
+  gpr_free(host);
+  gpr_free(port);
+  return s;
+}
+
+static std::unordered_map<string, std::deque<int>> get_hosts_and_cores(
+    const deque<string>& workers) {
+  std::unordered_map<string, std::deque<int>> hosts;
+  for (auto it = workers.begin(); it != workers.end(); it++) {
+    string host = get_host(*it);
```
Comment (dgquintas) [inline test/cpp/qps/driver.cc]: const

File: test/cpp/qps/driver.cc [hunk 6]
```diff
@@ -105,18 +141,18 @@ struct ClientData {
...
+    const ServerConfig& initial_server_config, size_t num_servers,
+    int warmup_seconds, int benchmark_seconds, int spawn_local_worker_count) {
   // ClientContext allocations (all are destroyed at scope exit)
   list<ClientContext> contexts;
 
   // To be added to the result, containing the final configuration used for
   // client and config (including host, etc.)
   ClientConfig result_client_config;
-  ServerConfig result_server_config;
+  ServerConfig result_server_config = initial_server_config;
```
Comment (dgquintas) [inline test/cpp/qps/driver.cc]: const

File: test/cpp/qps/driver.cc [hunk 7]
```diff
@@ -172,18 +211,49 @@ std::unique_ptr<ScenarioResult> RunScenario(
...
+    char* cli_target;
+    gpr_split_host_port(workers[i].c_str(), &host, &driver_port);
+    string host_str(host);
+    int server_core_limit = initial_server_config.core_limit();
+    int client_core_limit = initial_client_config.core_limit();
+
+    if (server_core_limit == 0 && client_core_limit > 0) {
+      // In this case, limit the server cores if it matches the
+      // same host as one or more clients
+      const auto& dq = hosts_cores[host_str];
```
Comment (dgquintas) [inline test/cpp/qps/driver.cc]: If `host_str` is always expected to be in `hosts_cores`, best use the `at()` method.

File: test/cpp/qps/driver.cc [hunk 8]
```diff
@@ -191,19 +261,50 @@ std::unique_ptr<ScenarioResult> RunScenario(
...
+            i + num_servers);
     clients[i].stub = WorkerService::NewStub(
-        CreateChannel(workers[i + num_servers], InsecureChannelCredentials()));
+        CreateChannel(worker, InsecureChannelCredentials()));
+    ClientConfig per_client_config = client_config;
+
+    int server_core_limit = initial_server_config.core_limit();
+    int client_core_limit = initial_client_config.core_limit();
+    if ((server_core_limit > 0) || (client_core_limit > 0)) {
+      auto& dq = hosts_cores[get_host(worker)];
```
Comment (dgquintas) [inline test/cpp/qps/driver.cc]: Ditto. Use `at()` instead?

**Comment by vjpai** [comment]
I think the comments are all resolved now.

**Comment by dgquintas** [comment]
LGTM, waiting for green. Thank you for yielding to the Const Police.

**Comment by vjpai** [comment]
Does this look green enough for you?

**Comment by dgquintas** [comment]
Hmm it does look "500 Server Error" to me. Were the failures "acceptable" if you recall?

**Comment by vjpai** [comment]
The only problems were pre-existing timeouts in Python tests and C++ end-to-end test. Neither should have been affected by this PR because they were not using any of the same code or protos, nor does this PR affect any of the underlying gRPC src or include.

**Pull request merged**